src/main/java/org/opensearch/knn/jni/JNIService.java

```diff
@@ -145,7 +192,21 @@ public static void free(long indexPointer, String engineName) {
             return;
         }
 
-        throw new IllegalArgumentException("Free not supported for provided engine");
+        throw new IllegalArgumentException(String.format("Free not supported for provided engine : %s", engineName));
+    }
+
+    /**
+     * Deallocate memory of the shared index state
+     *
+     * @param shareIndexStateAddr address of shared state
+     * @param engineName name of engine
+     */
+    public static void freeSharedIndexState(long shareIndexStateAddr, String engineName) {
```

Comment:
Thats an interesting point. I think JNIService is probably an unnecessary class. We could move the logic to the Engine and then the given Engine calls the proper JNI method (FaissService or NmslibService).

But, for this, I think we would want to refactor the whole JNIService class so I think it would be out of scope of this PR.

Reply:
It'll make the PR touch a good amount of unrelated files, but remove a lot of unnecessary method calls. I can update.

Reply:
See what works here. If we want to take that as a part of separate PR that is also fine.

Reply:
@navneet1v updated